### ISSUE TITLE
pyodide.http.FetchResponse throws exceptions for some requests based on the status code of the response

### ISSUE THREAD

**Issue by owenlamont** [pyodide/pyodide#3768]
## 🐛 Bug

pyodide.http.FetchResponse throws OSError exception for some response status codes (particularly 400 response - I'm not sure which others) instead of returning them.

### To Reproduce

```python
        <py-script>
            import asyncio
            import js
            import pyodide.http as http

            async def test_request():
                try:
                    response = await http.pyfetch(
                        "http://localhost:8080/end_point_that_returns_400_status"
                    )
                    if response.status == 400:
                        # I expected this to get hit but it doesn't
                        error_msg = await response.string()
                        js.alert(f"Handled 400 response with error message: {error_msg}")
                except Exception as e:
                    # This appears to be catching an OSError exception for the 400 response instead
                    js.alert(e)
            asyncio.ensure_future(test_request())
        </py-script>
```

I received an alert from the exception handler like this:

![image](https://user-images.githubusercontent.com/12672027/231931745-57f7d6f9-89de-43b9-9a79-fea0d57b2383.png)

### Expected behavior

I expected pyodide.http.FetchResponse to return the response regardless of what status code it had.

### Environment

- Pyodide Version: 22.1.0
- Browser version:  Vivaldi 5.7.2921.65 (Stable channel) (64-bit) 
- Any other relevant information: Using PyScript 2023.03.1 which I believe uses Pyodide 22.1.0 under the hood

### Additional context

I'm not sure if this is expected behaviour or not, but the [pyodide.http.FetchResponse documentation](https://pyodide.org/en/stable/usage/api/python-api/http.html#pyodide.http.FetchResponse) didn't warn that any response statuses could throw exceptions so if this behaviour is expected I think at the very least it is a documentation problem that should be addressed.

For context I was working on the front and backend of a system and I deliberately had the backend return a 400 response if the user failed to set some parameters correctly on the front end and I expected to be able to handle that response and communicate it to the user. Given the response instead throws an exception I don't think I can retrieve the additional information in the body of the response which I had planned to use.

Perhaps naively I expected pyodide.http.FetchResponse to work like a httpx async get request (I'm quite new to PyScript and pyodide). If this behaviour is actually due to some PyScript issue my apologies, I'll re-raise on their repo.

**Comment by ryanking13** [comment]
Hmm, I cannot reproduce the issue with 0.22.1. pyfetch should not throw exceptions on 4xx 5xx responses.

Could you check what happens if you do JS `fetch` in the browser console with the same URL? It is worth noting that pyfetch is a simple wrapper of [JS fetch](https://github.com/pyodide/pyodide/blob/4a0ca4ed48ff109679df4e246e8ed79901099f37/src/py/pyodide/http.py#L251) so the behavior should be basically the same.

**Comment by owenlamont** [comment]
Thanks for double checking... Maybe it is some logic in the PyScript wrapper. I'll try testing some more too. Happy for you to close this issue if it can't be reproduced.

**Comment by hoodmane** [comment]
I don't think this is a problem with PyScript or with Pyodide. Per the whatwg fetch spec, the JavaScript `fetch` function sometimes throws a TypeError. In these cases `pyfetch` throws an `OSError`. This is the intended behavior.

**Comment by owenlamont** [comment]
Thanks for clarifying that @hoodmane - apologies for raising this as a bug. I still think maybe a documentation update should happen to clarify when this should happen as I don't think it is obvious behavior.

**Comment by hoodmane** [comment]
> maybe a documentation update should happen to clarify when this should happen

Fair enough. It is sort of mentioned here: https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules

We could use to add an faq entry and to somehow hack the error message at runtime to explain what's going on.

**Comment by owenlamont** [comment]
For me I'd be happy with some tweaking of the [FetchResponse doc](https://pyodide.org/en/stable/usage/api/python-api/http.html#pyodide.http.FetchResponse) to state which statuses will return and which will raise exceptions.

**Comment by hoodmane** [comment]
Well I don't believe it's not that certain response codes raise errors. In this case I think the error is due to some other problem? I'm not fully sure to be honest.

What happens if you do
```
from js import fetch

await fetch(url);
```

**Comment by hoodmane** [comment]
> yes this issue as described was a mistake

Glad to hear you figured it out.

> I did have my server returning a plain/text body describing the cause of the error which I had hoped to capture on the client side

Hmm maybe you could open a separate issue specifically for this? If you provide a reproducer for the error that I can run, I can investigate for you.
